File: client/javascript/src/mpc/service.ts
/*
 * Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
 * TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 *
 *
 */

import {Invocation, Invoker} from "@/mpc/invoker";
import {Context} from "@/prsim";
import {composite, PROVIDER} from "@/mpc/filter";

class ServiceInvoker implements Invoker<any> {

    private readonly service: any

    constructor(service: any) {
        this.service = service;
    }

    run(ctx: Context, invocation: Invocation): Promise<any> {
        return new Promise<any>((resolve, reject) => {
            try {
                const args: any[] = [ctx, ...invocation.getArguments(ctx)];
                const ret = invocation.getInspector(ctx).invoke(this.service, args)
                if (ret instanceof Promise) {
                    ret.then(r => resolve(r)).catch(e => reject(e));
                    return;
                }
                resolve(ret);
            } catch (e) {
                reject(e)
            }
        })
    }

}

export class ServiceInvokeHandler implements Invoker<any> {

    private invoker: Invoker<any>;

    constructor(service: any) {
        this.invoker = composite(new ServiceInvoker(service), PROVIDER);
    }

    run(ctx: Context, invocation: Invocation): Promise<any> {
        return this.invoker.run(ctx, invocation)
    }

}

File: client/python/mesh/mpc/service.py
#
# Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
# TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
#
#
from asyncio import Future
from typing import Any

import mesh.log as log
from mesh.context import Mesh
from mesh.macro import T, ServiceLoader
from mesh.mpc.eden import Eden
from mesh.mpc.filter import Filter, PROVIDER
from mesh.mpc.invoker import Invoker, Invocation


class ServiceInvoker(Invoker[T]):

    def __init__(self, service: Any):
        self.service = service

    def run(self, invocation: Invocation) -> Any:
        result = invocation.get_inspector().invoke(self.service, invocation.get_arguments())
        if not isinstance(result, Future):
            return result
        try:
            eden = ServiceLoader.load(Eden).get_default()
            execution = eden.infer(Mesh.context().get_urn())
            timeout = execution.schema().timeout
            return result.result(timeout=max(timeout, 3000))
        except BaseException as e:
            log.error(f'Invoke {Mesh.context().get_urn()} fault because of {e}')
            raise e


class ServiceInvokeHandler(Invoker):
    """

    """

    def __init__(self, service: Any):
        self.invoker = Filter.composite(ServiceInvoker(service), PROVIDER)

    def run(self, invocation: Invocation) -> Any:
        return self.invoker.run(invocation)
